File: bot/models/perps.py
"""Perpetual trading models for HyperLiquid integration."""

from sqlalchemy import Column, Integer, String, DateTime, Boolean, ForeignKey, Text, Float, Numeric
from datetime import datetime
from database.db import Base


class PerpPosition(Base):
    """Perpetual trading position."""
    __tablename__ = "perp_positions"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    wallet_id = Column(Integer, ForeignKey("wallets.id"), nullable=True)

    exchange = Column(String(50), default="hyperliquid")
    market = Column(String(50), nullable=False)  # "ETH-USD", "BTC-USD"
    side = Column(String(10), nullable=False)  # "long" or "short"

    size = Column(Numeric(precision=20, scale=8), nullable=False)  # Position size
    entry_price = Column(Numeric(precision=20, scale=8), nullable=False)
    mark_price = Column(Numeric(precision=20, scale=8), nullable=True)
    leverage = Column(Integer, default=1)
    margin = Column(Numeric(precision=20, scale=8), nullable=True)

    unrealized_pnl = Column(Numeric(precision=20, scale=8), default=0)
    liquidation_price = Column(Numeric(precision=20, scale=8), nullable=True)

    tp_price = Column(Numeric(precision=20, scale=8), nullable=True)  # Take profit
    sl_price = Column(Numeric(precision=20, scale=8), nullable=True)  # Stop loss

    status = Column(String(20), default="open", index=True)  # open, closed, liquidated

    opened_at = Column(DateTime, default=datetime.utcnow)
    closed_at = Column(DateTime, nullable=True)
    closed_pnl = Column(Numeric(precision=20, scale=8), nullable=True)


class PerpOrder(Base):
    """Perpetual trading order."""
    __tablename__ = "perp_orders"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    position_id = Column(Integer, ForeignKey("perp_positions.id"), nullable=True)

    exchange = Column(String(50), default="hyperliquid")
    market = Column(String(50), nullable=False)
    side = Column(String(10), nullable=False)  # "long" or "short"

    order_type = Column(String(20), nullable=False)  # market, limit, stop_market, take_profit, stop_loss
    size = Column(Numeric(precision=20, scale=8), nullable=False)
    price = Column(Numeric(precision=20, scale=8), nullable=True)  # NULL for market orders
    leverage = Column(Integer, default=1)

    status = Column(String(20), default="pending", index=True)  # pending, filled, cancelled, failed
    hl_order_id = Column(String(100), nullable=True)  # HyperLiquid order ID

    fill_price = Column(Numeric(precision=20, scale=8), nullable=True)
    filled_at = Column(DateTime, nullable=True)

    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class HyperLiquidAccount(Base):
    """User's HyperLiquid account configuration."""
    __tablename__ = "hyperliquid_accounts"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, unique=True, index=True)

    hl_address = Column(String(255), nullable=True)  # HyperLiquid address
    api_key_encrypted = Column(Text, nullable=True)
    api_secret_encrypted = Column(Text, nullable=True)

    is_active = Column(Boolean, default=True)
    total_equity = Column(Numeric(precision=20, scale=8), default=0)
    available_margin = Column(Numeric(precision=20, scale=8), default=0)

    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

File: api-ts/src/db/schema/perps.ts
import {
	boolean,
	integer,
	numeric,
	pgTable,
	serial,
	timestamp,
	text,
	varchar,
} from 'drizzle-orm/pg-core'

export const perpPositions = pgTable('perp_positions', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull(),
	walletId: integer('wallet_id'),
	exchange: varchar('exchange', { length: 50 }).default('hyperliquid'),
	market: varchar('market', { length: 50 }).notNull(),
	side: varchar('side', { length: 10 }).notNull(),
	size: numeric('size', { precision: 20, scale: 8 }).notNull(),
	entryPrice: numeric('entry_price', { precision: 20, scale: 8 }).notNull(),
	markPrice: numeric('mark_price', { precision: 20, scale: 8 }),
	leverage: integer('leverage').default(1),
	margin: numeric('margin', { precision: 20, scale: 8 }),
	unrealizedPnl: numeric('unrealized_pnl', { precision: 20, scale: 8 }).default('0'),
	liquidationPrice: numeric('liquidation_price', { precision: 20, scale: 8 }),
	tpPrice: numeric('tp_price', { precision: 20, scale: 8 }),
	slPrice: numeric('sl_price', { precision: 20, scale: 8 }),
	status: varchar('status', { length: 20 }).default('open'),
	openedAt: timestamp('opened_at').defaultNow(),
	closedAt: timestamp('closed_at'),
	closedPnl: numeric('closed_pnl', { precision: 20, scale: 8 }),
})

export const perpOrders = pgTable('perp_orders', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull(),
	positionId: integer('position_id'),
	exchange: varchar('exchange', { length: 50 }).default('hyperliquid'),
	market: varchar('market', { length: 50 }).notNull(),
	side: varchar('side', { length: 10 }).notNull(),
	orderType: varchar('order_type', { length: 20 }).notNull(),
	size: numeric('size', { precision: 20, scale: 8 }).notNull(),
	price: numeric('price', { precision: 20, scale: 8 }),
	leverage: integer('leverage').default(1),
	status: varchar('status', { length: 20 }).default('pending'),
	hlOrderId: varchar('hl_order_id', { length: 100 }),
	fillPrice: numeric('fill_price', { precision: 20, scale: 8 }),
	filledAt: timestamp('filled_at'),
	createdAt: timestamp('created_at').defaultNow(),
	updatedAt: timestamp('updated_at').defaultNow(),
})

export const hyperliquidAccounts = pgTable('hyperliquid_accounts', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull().unique(),
	hlAddress: varchar('hl_address', { length: 255 }),
	apiKeyEncrypted: text('api_key_encrypted'),
	apiSecretEncrypted: text('api_secret_encrypted'),
	isActive: boolean('is_active').default(true),
	totalEquity: numeric('total_equity', { precision: 20, scale: 8 }).default('0'),
	availableMargin: numeric('available_margin', { precision: 20, scale: 8 }).default('0'),
	createdAt: timestamp('created_at').defaultNow(),
	updatedAt: timestamp('updated_at').defaultNow(),
})

export type PerpPosition = typeof perpPositions.$inferSelect
export type NewPerpPosition = typeof perpPositions.$inferInsert
export type PerpOrder = typeof perpOrders.$inferSelect
export type NewPerpOrder = typeof perpOrders.$inferInsert
export type HyperliquidAccount = typeof hyperliquidAccounts.$inferSelect
export type NewHyperliquidAccount = typeof hyperliquidAccounts.$inferInsert
